### PR TITLE
Append timestamp to autosave

### PR DESCRIPTION
**Pull request by atmaxinger**
Append a timestamp to the autosave. 

Please don’t merge yet as I want to add an option to the config.

### CODE REVIEW

**Comment by AaronVanGeffen** [comment]
Without any file management, this change would make the amount of saves pile up really quickly. Maybe you can prepend a sequence number to the date so you can retain a certain (configurable) amount of autosaves?

**Comment by atmaxinger** [comment]
Yeah, I plan to do that:
1.) No Timestamp
2.) Timestamp
3.) Sequence Number

**Comment by AaronVanGeffen** [comment]
My point is that either of the new options is useless without managing the files themselves. I wouldn't use either option if it means I get tens of new files after an hour of playing, while I only really need, say, the most recent five.

**Comment by Gymnasiast** [comment]
I don't think this warrants an option. Just keep the last ten files and use timestamps.

**Comment by atmaxinger** [comment]
Ok, will do that.

**Comment by atmaxinger** [comment]
I've implemented the deletion of old autosaves in my code. 

Unfortunately platform_delete_file is still a stub on posix systems :(

File: src/game.c [hunk 1]
```diff
@@ -1059,16 +1060,77 @@ void save_game_as()
...
+	{
+		return;
+	}
+	
+	// calculate how much saves we need to delete.
+	numAutosavesToDelete = autosavesCount - numberOfFilesToKeep;
+	
+	fileEnumHandle = platform_enumerate_files_begin(filter);
+	while (platform_enumerate_files_next(fileEnumHandle, &fileInfo)) {
+		if(numAutosavesToDelete > 0) { // we still have to delte some save files.
```
Comment (IntelOrca) [inline src/game.c]: Typo: delte

File: src/game.c [hunk 2]
```diff
@@ -1059,16 +1060,77 @@ void save_game_as()
...
-
+	utf8 timeString[21]="";
+	
+	time_t rawtime;
+	struct tm * timeinfo;
+	
+	time ( &rawtime );
+	timeinfo = localtime ( &rawtime );
+
+	delete_old_autosaves(9);
```
Comment (IntelOrca) [inline src/game.c]: Perhaps move this 9 to a constant at the top of the file, use #define or enum.

File: src/game.c [hunk 3]
```diff
@@ -1059,16 +1060,77 @@ void save_game_as()
...
+	strncat(filter, ".sv6", sizeof(filter) - strnlen(filter, MAX_PATH) - 1);
+	
+	// At first, count how many autosaves there are
+	fileEnumHandle = platform_enumerate_files_begin(filter);
+	while (platform_enumerate_files_next(fileEnumHandle, &fileInfo)) {
+		autosavesCount++;
+	}
+	platform_enumerate_files_end(fileEnumHandle);
+	
+	// If there are less autosaves then maxAutosaves we don't need to delete anything;
```
Comment (janisozaur) [inline src/game.c]: typo: than

File: src/game.c [hunk 4]
```diff
@@ -1059,16 +1060,77 @@ void save_game_as()
...
+	if(autosavesCount < numberOfFilesToKeep)
+	{
+		return;
+	}
+	
+	// calculate how much saves we need to delete.
```
Comment (janisozaur) [inline src/game.c]: s/much/many/

File: src/game.c [hunk 5]
```diff
@@ -1059,16 +1060,77 @@ void save_game_as()
...
+	while (platform_enumerate_files_next(fileEnumHandle, &fileInfo)) {
+		if(numAutosavesToDelete > 0) { // we still have to delte some save files.
+			platform_get_user_directory(path, "save");
+			strcat(path, fileInfo.path);
+			
+			platform_file_delete(path);
+			numAutosavesToDelete--;
+		}
+		else {
+			// we have deleted all necesary saves
```
Comment (janisozaur) [inline src/game.c]: s/necesary/necessary/

**Comment by janisozaur** [comment]
Technically this should work, but I have some comments.

We already have some time-related code in `platform/`, although it's using `gmtime` rather than `localtime`, like you do, which perhaps makes more sense. Originally I remember doing that because Windows already had time implemented using its WinApi functions, but seeing that this not limited to POSIX, perhaps we should switch all of platform to `time.h`? @IntelOrca what's your stance on that?

I don't like the name of the function, especially with the kind of parameter it receives. When I first looked at it, I thought the parameter to `delete_old_autosaves` is a count of how many saves to remove. Can you come up with a better name? Something like `limit_save_count`. Again @IntelOrca any comments?

One thing that is overlooked entirely, I think, is that with autosaves you generally have them in some order, and when limiting their amount, the user would like to get rid of the oldest first. I believe the platform file enumeration gives no guarantees on the order the files are enumerated in, but it rather depends on filesystem. That said, we can end up deleting most recent save every time we call this function and retaining old and long forgotten saves, which in my opinion renders the whole autosave functionality useless.

File: src/game.c [hunk 6]
```diff
@@ -1059,16 +1060,77 @@ void save_game_as()
...
+	int numAutosavesToDelete = 0;
+	int fileEnumHandle = 0;
+	int autosavesCount = 0;
+	file_info fileInfo;
+	
+	utf8 filter[MAX_PATH];
+	utf8 path[MAX_PATH];
+	
+	platform_get_user_directory(filter, "save");
+	strncat(filter, "autosave_*", sizeof(filter) - strnlen(filter, MAX_PATH) - 1);
```
Comment (janisozaur) [inline src/game.c]: There's no point in splitting this in two.

**Comment by IntelOrca** [comment]
* Regarding time, Windows returns filesystem times in 100-nanosecond precision from a certain point in time, I think posix is just seconds from a certain point in time. It might be good to be consistent and maybe have our own struct / wrappers.

* Yes, better naming is needed.

* Windows enumerates files in name order usually. We have timestamps available so you can delete the oldest modified file. Alternatively you might want to parse the name of the file to get the time.

**Comment by janisozaur** [comment]
https://msdn.microsoft.com/en-us/library/windows/desktop/aa364428%28v=vs.85%29.aspx

> The order in which the search returns the files, such as alphabetical order, is not guaranteed, and is dependent on the file system. If the data must be sorted, the application must do the ordering after obtaining all the results.

**Comment by IntelOrca** [comment]
> usually

File: src/game.c (hunk 2 above)
Comment (janisozaur) [inline src/game.c]: how about config entry?

Comment (IntelOrca) [inline src/game.c]: @Gymnasiast 
> I don't think this warrants an option. Just keep the last ten files and use timestamps.

File: src/game.c (hunk 3 above)
Comment (Gymnasiast) [inline src/game.c]: Also: "fewer autosaves", not "less" (this is nitpicky, I know).

File: src/game.c [hunk 7]
```diff
@@ -1059,16 +1062,97 @@ void save_game_as()
 	window_loadsave_open(LOADSAVETYPE_SAVE | LOADSAVETYPE_GAME, (char*)path_get_filename(gScenarioSavePath));
 }
 
+int compare_autosave_file_paths (const void * a, const void * b ) {
+	return strcmp(*(char **)a, *(char **)b);
+}
+
+void limit_autosave_count(const int numberOfFilesToKeep)
```
Comment (janisozaur) [inline src/game.c]: the argument here should of `size_t`

File: src/game.c [hunk 8]
```diff
@@ -1059,16 +1062,97 @@ void save_game_as()
...
+{
+	int numAutosavesToDelete = 0;
+	int fileEnumHandle = 0;
+	int autosavesCount = 0;
```
Comment (janisozaur) [inline src/game.c]: `size_t`

File: src/game.c [hunk 9]
```diff
@@ -1059,16 +1062,97 @@ void save_game_as()
...
+	file_info fileInfo;
+	
+	utf8 filter[MAX_PATH];
+	
+	utf8 **autosaveFiles = NULL;
+	
+	int i=0;
```
Comment (janisozaur) [inline src/game.c]: `size_t`

File: src/game.c [hunk 10]
```diff
@@ -1059,16 +1062,97 @@ void save_game_as()
...
+	
+	// At first, count how many autosaves there are
+	fileEnumHandle = platform_enumerate_files_begin(filter);
+	while (platform_enumerate_files_next(fileEnumHandle, &fileInfo)) {
+		autosavesCount++;
+	}
+	platform_enumerate_files_end(fileEnumHandle);
+	
+	// If there are fewer autosaves than maxAutosaves we don't need to delete anything;
+	if(autosavesCount < numberOfFilesToKeep) {
```
Comment (janisozaur) [inline src/game.c]: This should be checking `<=` condition

**Comment by janisozaur** [comment]
Only minor comments now, mostly w.r.t. signedness of variables.

Otherwise :+1: from me.

File: src/game.c (hunk 10 above)
Comment (atmaxinger) [inline src/game.c]: whoops :D

**Comment by janisozaur** [comment]
how can `numAutosavesToDelete` be negative?

**Comment by atmaxinger** [comment]
Oh well, it can't. I shouldn't do midnight coding ...

**Comment by Broxzier** [comment]
You can use `git add` to stage changes, and use `git commit` when you finished something.

**Comment by atmaxinger** [comment]
Is there anything left to do?

**Comment by janisozaur** [comment]
I'm happy with the code, though personally I'd ask you to squash the history so it's a bit cleaner.

**Comment by Gymnasiast** [comment]
:+1:

**Pull request merged**